### PR TITLE
When sanitizing html, disable link creation when already under an anc…

### PR DESCRIPTION
**Pull request by hagenrd**
…hor.

We were running into this:
```html
<p><a data-nti-entity-type="LINK" data-nti-entity-mutability="MUTABLE" href="http://www.google.com" data-nti-entity-href="http://www.google.com" data-nti-entity-contiguous="false" data-nti-entity-autoLink="true" data-nti-entity-has-preview="preview">www.google.com</a></p>
```

getting turned into this:
```html
<html><body><p><a data-nti-entity-mutability="MUTABLE" data-nti-entity-type="LINK" href="http://www.google.com"></a><a href="http://www.google.com">www.google.com</a></p></body></html>
```

### CODE REVIEW

File: src/nti/contentfragments/html.py [hunk 1]
```diff
@@ -233,6 +235,13 @@ def sanitize_token(self, token):
             # character data beneath a rejected element
             return None
 
+        # Indicate whether we're in an anchor tag
+        if token.get('name') == 'a':
+            if token_type == 'StartTag':
+                self._in_anchor = True
+            elif token_type == 'EndTag':
+                self._in_anchor = False
```
Comment (jamadden) [inline src/nti/contentfragments/html.py]: I was tempted to shorten this to just `self._in_anchor = token_type == 'StartTag'` and it took a bit to dawn on me why that's not right. Maybe a quick comment about why that's no good?

File: src/nti/contentfragments/tests/test_html.py [hunk 2]
```diff
@@ -176,6 +176,12 @@ def test_allowed_attribute_provider(self):
             for attr_name in allowed_attrs:
                 self._test_allowed_attribute_provider(attr_name)
 
+    def test_existing_links(self):
+        # Ensure we properly handle html with existing anchors
+        html = '<html><body><p><a href="http://www.google.com">www.google.com</a></p></body></html>'
```
Comment (jamadden) [inline src/nti/contentfragments/tests/test_html.py]: I'd also like this to have a link in a non-standard 'data' attribute, like in the motivating example.

**Review by jamadden — COMMENTED** [review]
Needs a good entry in CHANGES.rst, and possibly a little test addition, but other than that (and one minor, take-it-or-leave-it suggestion), LPGTM.

I do wonder what happens with nested `<a>` tags (whether that happens in practice or not)...

File: src/nti/contentfragments/tests/test_html.py (hunk 2 above)
Comment (hagenrd) [reply src/nti/contentfragments/tests/test_html.py]: If by that you mean an additional attribute like `data-nti-entity-href="http://www.google.com"`, that'll just get stripped out.  I suppose I could add it to an allowed attribute provider, if that's what you were referring to.

Comment (jamadden) [reply src/nti/contentfragments/tests/test_html.py]: I do mean exactly that. The point being to demonstrate that the 'find links in text' stuff isn't happening there and expanding in a weird, unexpected place. I could imagine that if it weren't stripped,  etc, this might not be sufficient. Defense against future changes.

Comment (hagenrd) [reply src/nti/contentfragments/tests/test_html.py]: Fair enough.

File: src/nti/contentfragments/html.py (hunk 1 above)
Comment (hagenrd) [reply src/nti/contentfragments/html.py]: Huh, well I'd assumed EndTag would cause issues, but I don't seem to be able to cause issues with it, as an empty anchor tag with url-like text following seems to get turned into a starttag-characters-endtag sequence, e.g. `<a href="http://www.google.com"/>www.google.com` effectively becomes `<a href="http://www.google.com">www.google.com</a>`.  Is there another reason to avoid that?

**Comment by hagenrd** [comment]
> I do wonder what happens with nested <a> tags (whether that happens in practice or not)...

There seems to be some indication that's not legal syntax, from the text-level semantics of the anchor element at [W3C](https://www.w3.org/TR/html52/textlevel-semantics.html#the-a-element):
> Content Model:
> Transparent, but there must be no interactive content or a element descendants.

File: src/nti/contentfragments/html.py (hunk 1 above)
Comment (jamadden) [reply src/nti/contentfragments/html.py]: If you ever get another token type (that has a `name`) besides `StartTag` or `EndTag`, then the shortcut syntax would be broken (I can't remember what the token types are or whether their data ever includes 'name'; I suspect most readers of the code won't remember either).

**Comment by jamadden** [comment]
>> I do wonder what happens with nested tags (whether that happens in practice or not)…

> There seems to be some indication that's not legal syntax,

True enough, but it doesn't mean it never happens. `<a>` can have complex descendants, like this:
```html
<a class="js-selected-navigation-item UnderlineNav-item hx_underlinenav-item no-wrap js-responsive-underlinenav-item" data-tab-item="issues-tab" data-hotkey="g i" data-ga-click="Repository, Navigation click, Issues tab" data-selected-links="repo_issues repo_labels repo_milestones /NextThought/nti.contentfragments/issues" href="/NextThought/nti.contentfragments/issues">
      <svg height="16" class="octicon octicon-issue-opened UnderlineNav-octicon d-none d-sm-inline" classes="UnderlineNav-octicon" display="none inline" viewBox="0 0 16 16" version="1.1" width="16" aria-hidden="true">
        <path fill-rule="evenodd" d="M8 1.5a6.5 6.5 0 100 13 6.5 6.5 0 000-13zM0 8a8 8 0 1116 0A8 8 0 010 8zm9 3a1 1 0 11-2 0 1 1 0 012 0zm-.25-6.25a.75.75 0 00-1.5 0v3.5a.75.75 0 001.5 0v-3.5z"></path>
      </svg>

      <span data-content="Issues">Issues</span>
      <span class="Counter " title="0" hidden="hidden">0</span>
 </a>
```

I can easily imagine certain ways of putting pages together (e.g., components) producing nested `a` tags; perhaps just serving as page fragment anchors even. I don't know what the parser would do with that or if it matters...

**Review by jzuech3 — APPROVED** [review]
LGTM

**Review by jamadden — APPROVED** [review]
A quick test of the nested anchor handling would be appreciated, otherwise LGTM.

**Comment by hagenrd** [comment]
Unfortunately the parser seems to be splitting the nesting before our logic, so we don't actually see the nested anchors using the current parser.

**Comment by jamadden** [comment]
Excellent. Thanks!

**Comment by jamadden** [comment]
1.5.0 is on PyPI.

**Comment by hagenrd** [comment]
Thanks, Jason!
